pkgs/development/interpreters/python/mk-python-derivation.nix

```diff
@@ -57,9 +57,12 @@ python.stdenv.mkDerivation (builtins.removeAttrs attrs ["disabled"] // {
 
   inherit pythonPath;
 
-  # patch python interpreter to write null timestamps when compiling python files
-  # this way python doesn't try to update them when we freeze timestamps in nix store
+
+  # Determinism: The interpreter is patched to write null timestamps when compiling python files.
+  # This way python doesn't try to update them when we freeze timestamps in nix store.
   DETERMINISTIC_BUILD=1;
+  # Determinism: We fix the hashes of str, bytes and datetime objects.
+  PYTHONHASHSEED = 0;
```

Comment:
I disagree. What is `nix-shell` used or designed for? For testing/building a package, or for building environments with certain packages? In my point of view its designed for the former, but its (mis)used a lot for the latter as well. If we are going to make changes for `nix-shell` then we should also unset `SOURCE_DATE_EPOCH` and likely a dozen other vars as well which will render it useless for the former use case.

In Nix 1.12 we have `nix run` which, if I am correct, shouldn't set such env vars (unfortunately it still does).